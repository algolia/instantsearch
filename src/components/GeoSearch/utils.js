export const getBoundsZoomLevel = (bounds, mapDim) => {
  const WORLD_DIM = { height: 256, width: 256 };
  const ZOOM_MAX = 21;

  const latRad = lat => {
    const sin = Math.sin(lat * Math.PI / 180);
    const radX2 = Math.log((1 + sin) / (1 - sin)) / 2;

    return Math.max(Math.min(radX2, Math.PI), -Math.PI) / 2;
  };

  const zoom = (mapPx, worldPx, fraction) =>
    Math.floor(Math.log(mapPx / worldPx / fraction) / Math.LN2);

  const ne = bounds.getNorthEast();
  const sw = bounds.getSouthWest();

  const latFraction = (latRad(ne.lat()) - latRad(sw.lat())) / Math.PI;

  const lngDiff = ne.lng() - sw.lng();
  const lngFraction = (lngDiff < 0 ? lngDiff + 360 : lngDiff) / 360;

  const latZoom = zoom(mapDim.height, WORLD_DIM.height, latFraction);
  const lngZoom = zoom(mapDim.width, WORLD_DIM.width, lngFraction);

  return Math.min(latZoom, lngZoom, ZOOM_MAX);
};
