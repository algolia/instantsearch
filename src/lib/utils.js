import reduce from 'lodash/collection/reduce';
import forEach from 'lodash/collection/forEach';
import find from 'lodash/collection/find';
import get from 'lodash/object/get';
import isEmpty from 'lodash/lang/isEmpty';
import keys from 'lodash/object/keys';
import uniq from 'lodash/array/uniq';
import mapKeys from 'lodash/object/mapKeys';

export {
  getContainerNode,
  bemHelper,
  prepareTemplateProps,
  isSpecialClick,
  isDomElement,
  getRefinements,
  clearRefinementsFromState,
  clearRefinementsAndSearch,
  prefixKeys
};

/**
 * Return the container. If it's a string, it is considered a
 * css selector and retrieves the first matching element. Otherwise
 * test if it validates that it's a correct DOMElement.
 * @param {string|DOMElement} selectorOrHTMLElement a selector or a node
 * @return {DOMElement} The resolved DOMElement
 * @throws Error when the type is not correct
 */
function getContainerNode(selectorOrHTMLElement) {
  const isFromString = typeof selectorOrHTMLElement === 'string';
  let domElement;
  if (isFromString) {
    domElement = document.querySelector(selectorOrHTMLElement);
  } else {
    domElement = selectorOrHTMLElement;
  }

  if (!isDomElement(domElement)) {
    let errorMessage = 'Container must be `string` or `HTMLElement`.';
    if (isFromString) {
      errorMessage += ` Unable to find ${selectorOrHTMLElement}`;
    }
    throw new Error(errorMessage);
  }

  return domElement;
}

/**
 * Returns true if the parameter is a DOMElement.
 * @param {any} o the value to test
 * @return {boolean} true if o is a DOMElement
 */
function isDomElement(o) {
  return o instanceof window.HTMLElement || Boolean(o) && o.nodeType > 0;
}

function isSpecialClick(event) {
  const isMiddleClick = event.button === 1;
  return isMiddleClick || event.altKey || event.ctrlKey || event.metaKey || event.shiftKey;
}

/**
 * Creates BEM class name according the vanilla BEM style.
 * @param {string} block the main block
 * @return {function} function that takes up to 2 parameters
 * that determine the element and the modifier of the BEM class.
 */
function bemHelper(block) {
  return function(element, modifier) {
    // block--element
    if (element && !modifier) {
      return `${block}--${element}`;
    }
    // block--element__modifier
    if (element && modifier) {
      return `${block}--${element}__${modifier}`;
    }
    // block__modifier
    if (!element && modifier) {
      return `${block}__${modifier}`;
    }

    return block;
  };
}

/**
 * Prepares an object to be passed to the Template widget
 * @param {object} unknownBecauseES6 an object with the following attributes:
 *  - transformData
 *  - defaultTemplate
 *  - templates
 *  - templatesConfig
 * @return {object} the configuration with the attributes:
 *  - transformData
 *  - defaultTemplate
 *  - templates
 *  - useCustomCompileOptions
 */
function prepareTemplateProps({
  transformData,
  defaultTemplates,
  templates,
  templatesConfig
}) {
  const preparedTemplates = prepareTemplates(defaultTemplates, templates);

  return {
    transformData,
    templatesConfig,
    ...preparedTemplates
  };
}

function prepareTemplates(defaultTemplates = {}, templates = {}) {
  const allKeys = uniq([...(keys(defaultTemplates)), ...(keys(templates))]);

  return reduce(allKeys, (config, key) => {
    const defaultTemplate = defaultTemplates[key];
    const customTemplate = templates[key];
    const isCustomTemplate = customTemplate !== undefined && customTemplate !== defaultTemplate;

    config.templates[key] = isCustomTemplate ? customTemplate : defaultTemplate;
    config.useCustomCompileOptions[key] = isCustomTemplate;

    return config;
  }, {templates: {}, useCustomCompileOptions: {}});
}

function getRefinement(state, type, attributeName, name, resultsFacets) {
  const res = {type, attributeName, name};
  let facet = find(resultsFacets, {name: attributeName});
  let count;
  if (type === 'hierarchical') {
    const facetDeclaration = state.getHierarchicalFacetByName(attributeName);
    const splitted = name.split(facetDeclaration.separator);
    res.name = splitted[splitted.length - 1];
    for (let i = 0; facet !== undefined && i < splitted.length; ++i) {
      facet = find(facet.data, {name: splitted[i]});
    }
    count = get(facet, 'count');
  } else {
    count = get(facet, `data["${res.name}"]`);
  }
  const exhaustive = get(facet, 'exhaustive');
  if (count !== undefined) {
    res.count = count;
  }
  if (exhaustive !== undefined) {
    res.exhaustive = exhaustive;
  }
  return res;
}

function getRefinements(results, state) {
  const res = [];

  forEach(state.facetsRefinements, (refinements, attributeName) => {
    forEach(refinements, name => {
      res.push(getRefinement(state, 'facet', attributeName, name, results.facets));
    });
  });

  forEach(state.facetsExcludes, (refinements, attributeName) => {
    forEach(refinements, name => {
      res.push({type: 'exclude', attributeName, name, exclude: true});
    });
  });

  forEach(state.disjunctiveFacetsRefinements, (refinements, attributeName) => {
    forEach(refinements, name => {
      res.push(getRefinement(state, 'disjunctive', attributeName, name, results.disjunctiveFacets));
    });
  });

  forEach(state.hierarchicalFacetsRefinements, (refinements, attributeName) => {
    forEach(refinements, name => {
      res.push(getRefinement(state, 'hierarchical', attributeName, name, results.hierarchicalFacets));
    });
  });

  forEach(state.numericRefinements, (operators, attributeName) => {
    forEach(operators, (values, operator) => {
      forEach(values, value => {
        res.push({
          type: 'numeric',
          attributeName,
          name: `${value}`,
          numericValue: value,
          operator
        });
      });
    });
  });

  forEach(state.tagRefinements, name => {
    res.push({type: 'tag', attributeName: '_tags', name});
  });

  return res;
}

function clearRefinementsFromState(inputState, attributeNames) {
  let state = inputState;

  if (isEmpty(attributeNames)) {
    state = state.clearTags();
    state = state.clearRefinements();
    return state;
  }

  forEach(attributeNames, attributeName => {
    if (attributeName === '_tags') {
      state = state.clearTags();
    } else {
      state = state.clearRefinements(attributeName);
    }
  });

  return state;
}

function clearRefinementsAndSearch(helper, attributeNames) {
  helper.setState(clearRefinementsFromState(helper.state, attributeNames)).search();
}

function prefixKeys(prefix, obj) {
  if (obj) {
    return mapKeys(obj, (v, k) => prefix + k);
  }

  return undefined;
}
