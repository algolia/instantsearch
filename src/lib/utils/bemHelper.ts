/**
 * Creates BEM class name according the vanilla BEM style.
 */
function bemHelper(block: string) {
  return function(element: string, modifier: string) {
    // block--element
    if (element && !modifier) {
      return `${block}--${element}`;
    }

    // block--element__modifier
    if (element && modifier) {
      return `${block}--${element}__${modifier}`;
    }

    // block__modifier
    if (!element && modifier) {
      return `${block}__${modifier}`;
    }

    return block;
  };
}

export default bemHelper;
