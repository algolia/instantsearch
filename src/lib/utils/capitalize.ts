function capitalize(text: string): string {
  return (
    text
      .toString()
      .charAt(0)
      .toUpperCase() + text.toString().slice(1)
  );
}

export default capitalize;
