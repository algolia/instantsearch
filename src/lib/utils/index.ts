export { default as capitalize } from './capitalize';
export { default as defer } from './defer';
export { default as isDomElement } from './isDomElement';
export { default as getContainerNode } from './getContainerNode';
export { default as isSpecialClick } from './isSpecialClick';
export { default as prepareTemplateProps } from './prepareTemplateProps';
export { default as renderTemplate } from './renderTemplate';
export { default as getRefinements } from './getRefinements';
export { default as clearRefinements } from './clearRefinements';
export { default as escapeRefinement } from './escapeRefinement';
export { default as enhanceConfiguration } from './enhanceConfiguration';
export { default as unescapeRefinement } from './unescapeRefinement';
export { default as checkRendering } from './checkRendering';
export { default as getPropertyByPath } from './getPropertyByPath';
export { default as noop } from './noop';
export { default as isFiniteNumber } from './isFiniteNumber';
export { default as isPlainObject } from './isPlainObject';
export { default as uniq } from './uniq';
export { default as range } from './range';
export { default as isEqual } from './isEqual';
export { default as escape } from './escape';
export { default as find } from './find';
export { default as mergeDeep } from './mergeDeep';
export { default as resolveSearchParameters } from './resolveSearchParameters';
export { warning, deprecate } from './logger';
export {
  createDocumentationLink,
  createDocumentationMessageGenerator,
} from './documentation';
export {
  aroundLatLngToPosition,
  insideBoundingBoxToBoundingBox,
} from './geo-search';
export { addAbsolutePosition } from './hits-absolute-position';
export { addQueryID } from './hits-query-id';
