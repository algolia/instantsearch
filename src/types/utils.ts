/* eslint-disable-next-line @typescript-eslint/generic-type-naming */
export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
