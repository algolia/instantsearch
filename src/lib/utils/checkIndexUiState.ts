import capitalize from './capitalize';
import { warning } from './logger';
import { Index } from '../../widgets/index/index';
import { Widget, IndexUiState } from '../../types';

// Some connectors are responsible for multiple widgets so we need
// to map them.
function getWidgetNames(connectorName: string): string[] {
  switch (connectorName) {
    case 'range':
      return [];

    case 'menu':
      return ['menu', 'menuSelect'];

    default:
      return [connectorName];
  }
}

type StateDescription = {
  connectors: string[];
  widgets: Array<Widget['$$type']>;
};

type StateToWidgets = {
  [TParameter in keyof IndexUiState]: StateDescription;
};

type MissingWidgets = Array<[string, StateDescription]>;

const stateToWidgetsMap: StateToWidgets = {
  query: {
    connectors: ['connectSearchBox'],
    widgets: ['ais.searchBox', 'ais.autocomplete', 'ais.voiceSearch'],
  },
  refinementList: {
    connectors: ['connectRefinementList'],
    widgets: ['ais.refinementList'],
  },
  menu: {
    connectors: ['connectMenu'],
    widgets: ['ais.menu'],
  },
  hierarchicalMenu: {
    connectors: ['connectHierarchicalMenu'],
    widgets: ['ais.hierarchicalMenu'],
  },
  numericMenu: {
    connectors: ['connectNumericMenu'],
    widgets: ['ais.numericMenu'],
  },
  ratingMenu: {
    connectors: ['connectRatingMenu'],
    widgets: ['ais.ratingMenu'],
  },
  range: {
    connectors: ['connectRange'],
    widgets: ['ais.rangeInput', 'ais.rangeSlider', 'ais.range'],
  },
  toggle: {
    connectors: ['connectToggleRefinement'],
    widgets: ['ais.toggleRefinement'],
  },
  geoSearch: {
    connectors: ['connectGeoSearch'],
    widgets: ['ais.geoSearch'],
  },
  sortBy: {
    connectors: ['connectSortBy'],
    widgets: ['ais.sortBy'],
  },
  page: {
    connectors: ['connectPagination'],
    widgets: ['ais.pagination', 'ais.infiniteHits'],
  },
  hitsPerPage: {
    connectors: ['connectHitsPerPage'],
    widgets: ['ais.hitsPerPage'],
  },
  configure: {
    connectors: ['connectConfigure'],
    widgets: ['ais.configure'],
  },
  places: {
    connectors: [],
    widgets: ['ais.places'],
  },
};

type CheckIndexUiStateParams = {
  index: Index;
  indexUiState: IndexUiState;
};

export function checkIndexUiState({
  index,
  indexUiState,
}: CheckIndexUiStateParams) {
  const mountedWidgets = index
    .getWidgets()
    .map(widget => widget.$$type)
    .filter(Boolean);

  const missingWidgets = Object.keys(indexUiState).reduce<MissingWidgets>(
    (acc, parameter) => {
      const requiredWidgets: Array<Widget['$$type']> | undefined =
        stateToWidgetsMap[parameter] && stateToWidgetsMap[parameter].widgets;

      if (
        requiredWidgets &&
        !requiredWidgets.some(requiredWidget =>
          mountedWidgets.includes(requiredWidget)
        )
      ) {
        acc.push([
          parameter,
          {
            connectors: stateToWidgetsMap[parameter].connectors,
            widgets: stateToWidgetsMap[parameter].widgets.map(
              (widgetIdentifier: string) => widgetIdentifier.split('ais.')[1]
            ),
          },
        ]);
      }

      return acc;
    },
    []
  );

  warning(
    missingWidgets.length === 0,
    `The UI state for the index "${index.getIndexId()}" is not consistent with the widgets mounted.

This can happen when the UI state is specified via \`initialUiState\`, \`routing\` or \`setUiState\` but that the widgets responsible for this state were not added. This results in those query parameters not being sent to the API.

To fully reflect the state, some widgets need to be added to the index "${index.getIndexId()}":

${missingWidgets
  .map(([stateParameter, { widgets }]) => {
    return `- \`${stateParameter}\` needs one of these widgets: ${([] as string[])
      .concat(...widgets.map(name => getWidgetNames(name!)))
      .map((name: string) => `"${name}"`)
      .join(', ')}`;
  })
  .join('\n')}

If you do not wish to display widgets but still want to support their search parameters, you can mount "virtual widgets" that don't render anything:

\`\`\`
${missingWidgets
  .filter(([_stateParameter, { connectors }]) => {
    return connectors.length > 0;
  })
  .map(([_stateParameter, { connectors, widgets }]) => {
    const capitalizedWidget = capitalize(widgets[0]!);
    const connectorName = connectors[0];

    return `const virtual${capitalizedWidget} = ${connectorName}(() => null);`;
  })
  .join('\n')}

search.addWidgets([
  ${missingWidgets
    .filter(([_stateParameter, { connectors }]) => {
      return connectors.length > 0;
    })
    .map(([_stateParameter, { widgets }]) => {
      const capitalizedWidget = capitalize(widgets[0]!);

      return `virtual${capitalizedWidget}({ /* ... */ })`;
    })
    .join(',\n  ')}
]);
\`\`\`

If you're using custom widgets that do set these query parameters, we recommend using connectors instead.

See https://www.algolia.com/doc/guides/building-search-ui/widgets/customize-an-existing-widget/js/#customize-the-complete-ui-of-the-widgets`
  );
}
