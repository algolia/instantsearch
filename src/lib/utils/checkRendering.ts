import { Renderer } from '../../types/connector';

function checkRendering(rendering: Renderer<any>, usage: string) {
  if (rendering === undefined || typeof rendering !== 'function') {
    throw new Error(`The render function is not valid (got type "${typeof rendering}").

${usage}`);
  }
}

export default checkRendering;
