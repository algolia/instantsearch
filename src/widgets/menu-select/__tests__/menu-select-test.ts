import { render as preactRender, VNode } from 'preact';
import algoliasearchHelper, { SearchParameters } from 'algoliasearch-helper';
import menuSelect from '../menu-select';
import { createSearchClient } from '../../../../test/mock/createSearchClient';
import {
  createDisposeOptions,
  createInitOptions,
  createRenderOptions,
} from '../../../../test/mock/createWidget';
import { castToJestMock } from '../../../../test/utils/castToJestMock';

const render = castToJestMock(preactRender);
jest.mock('preact', () => {
  const module = jest.requireActual('preact');

  module.render = jest.fn();

  return module;
});

describe('menuSelect', () => {
  describe('Usage', () => {
    it('throws without container ', () => {
      expect(() => {
        // @ts-ignore
        menuSelect({ container: undefined });
      }).toThrowErrorMatchingInlineSnapshot(`
"The \`container\` option is required.

See documentation: https://www.algolia.com/doc/api-reference/widgets/menu-select/js/"
`);
    });
  });

  describe('Lifecycle', () => {
    let data;
    let results;
    let state;
    let helper;

    beforeEach(() => {
      data = { data: [{ name: 'foo' }, { name: 'bar' }] };
      results = { getFacetValues: jest.fn(() => data) };
      helper = algoliasearchHelper(createSearchClient(), 'index_name');
      helper.search = jest.fn();
      state = helper.state;

      render.mockClear();
    });

    describe('render', () => {
      it('renders correctly', () => {
        const widget = menuSelect({
          container: document.createElement('div'),
          attribute: 'test',
        });

        widget.init!(
          createInitOptions({
            helper,
            createURL: () => '#',
          })
        );
        widget.render!(
          createRenderOptions({ results, createURL: () => '#', state })
        );

        const firstRender = render.mock.calls[0][0] as VNode;

        expect(firstRender.props).toMatchSnapshot();
      });

      it('renders transformed items correctly', () => {
        const widget = menuSelect({
          container: document.createElement('div'),
          attribute: 'test',
          transformItems: items =>
            items.map(item => ({ ...item, transformed: true })),
        });

        widget.init!(
          createInitOptions({
            helper,
            createURL: () => '#',
          })
        );
        widget.render!(
          createRenderOptions({ results, createURL: () => '#', state })
        );

        const firstRender = render.mock.calls[0][0] as VNode;

        expect(firstRender.props).toMatchSnapshot();
      });
    });

    describe('dispose', () => {
      it('unmounts the component', () => {
        const container = document.createElement('div');
        const widget = menuSelect({
          attribute: 'test',
          container,
        });

        helper.setState(
          widget.getWidgetSearchParameters!(new SearchParameters({}), {
            uiState: {
              menu: {
                amazingBrand: 'algolia',
              },
            },
          })
        );

        expect(helper.state).toEqual(
          new SearchParameters({
            hierarchicalFacets: [{ attributes: ['test'], name: 'test' }],
            hierarchicalFacetsRefinements: { test: [] },
            maxValuesPerFacet: 10,
          })
        );

        expect(render).toHaveBeenCalledTimes(0);

        const newState = widget.dispose!(
          createDisposeOptions({
            state: helper.state,
            helper,
          })
        );

        expect(render).toHaveBeenCalledTimes(1);
        expect(render).toHaveBeenLastCalledWith(null, container);
        expect(newState).toEqual(new SearchParameters());
      });
    });
  });
});
