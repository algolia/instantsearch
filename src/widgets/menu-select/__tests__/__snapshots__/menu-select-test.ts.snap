// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`menuSelect Lifecycle render renders correctly 1`] = `
{
  "cssClasses": {
    "noRefinementRoot": "ais-MenuSelect--noRefinement",
    "option": "ais-MenuSelect-option",
    "root": "ais-MenuSelect",
    "select": "ais-MenuSelect-select",
  },
  "items": [
    {
      "label": "foo",
      "value": undefined,
    },
    {
      "label": "bar",
      "value": undefined,
    },
  ],
  "refine": [Function],
  "templateProps": {
    "templates": {
      "defaultOption": "See all",
      "item": "{{label}} ({{#helpers.formatNumber}}{{count}}{{/helpers.formatNumber}})",
    },
    "templatesConfig": {},
    "useCustomCompileOptions": {
      "defaultOption": false,
      "item": false,
    },
  },
}
`;

exports[`menuSelect Lifecycle render renders transformed items correctly 1`] = `
{
  "cssClasses": {
    "noRefinementRoot": "ais-MenuSelect--noRefinement",
    "option": "ais-MenuSelect-option",
    "root": "ais-MenuSelect",
    "select": "ais-MenuSelect-select",
  },
  "items": [
    {
      "label": "foo",
      "transformed": true,
      "value": undefined,
    },
    {
      "label": "bar",
      "transformed": true,
      "value": undefined,
    },
  ],
  "refine": [Function],
  "templateProps": {
    "templates": {
      "defaultOption": "See all",
      "item": "{{label}} ({{#helpers.formatNumber}}{{count}}{{/helpers.formatNumber}})",
    },
    "templatesConfig": {},
    "useCustomCompileOptions": {
      "defaultOption": false,
      "item": false,
    },
  },
}
`;
